Add support for Ravnur Media Services

vstratovych opened the pull request:
## Purpose
This PR adds possibility to connect to RMS instance Ravnur Media Services is one of the possible AMS replacements.
Thanks to identical public API it's possible to use the same SDK and easily switch existing application from AMS to RMS.

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[x] User Interface
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* Smoke test that all basic features are working

## Other Information
Detailed documentation about Ravnur Media Services can be found here:
https://github.com/Ravnur-Inc/ams-api-replacement-demo-app

vstratovych committed:
Add support for Ravnur Media Services

Implemented configuration and client for RMS connection
Extended login form with new option "Connect to RMS account"

vstratovych committed:
Adjust UI for Ravnur Media Services connection

Conditionally show/hide features available only in AMS connection
Fixed some minor issues with nullable values handling

dima-ravnur commented on AMSExplorer/AMSClient/AMSClientV3.cs:
```diff
@@ -28,6 +23,16 @@
+
+using Microsoft.Identity.Client;
+using Microsoft.Rest;
+
```
Lets don't touch things which aren't related to our change

dima-ravnur commented on AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs:
```diff
@@ -88,7 +88,7 @@ public async Task InitAsync(AMSClientV3 amsClient)
                     Id = o.Data.Id,
                     Description = o.Data.Description,
-                    CDN = ((bool)o.Data.IsCdnEnabled) ? StreamingEndpointInformation.ReturnDisplayedProvider(o.Data.CdnProvider) ?? "CDN" : string.Empty,
+                    CDN = ((bool)o.Data.IsCdnEnabled.GetValueOrDefault()) ? StreamingEndpointInformation.ReturnDisplayedProvider(o.Data.CdnProvider) ?? "CDN" : string.Empty,
```
If this field is missing on our side. I'd suggest to add it to our response and do not change this code

vstratovych replied:
Yes, the value of the field returned from our API is null.

This is totally acceptable according to the nullable value in the model, so this is actually issue in the AMS Explorer code.

However, I agree that we can make change on our side in order to simplify overall complexity of this PR.

dima-ravnur commented on AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs:
```diff
@@ -133,7 +133,7 @@ public async Task RefreshStreamingEndpointAsync(StreamingEndpointResource stream
                     _MyObservStreamingEndpoints[index].LastModifiedOn = streamingEndpoint.Data.LastModifiedOn?.DateTime.ToLocalTime();
                     _MyObservStreamingEndpoints[index].Type = StreamingEndpointInformation.ReturnTypeSE(streamingEndpoint);
-                    _MyObservStreamingEndpoints[index].CDN = ((bool)streamingEndpoint.Data.IsCdnEnabled) ? StreamingEndpointInformation.ReturnDisplayedProvider(streamingEndpoint.Data.CdnProvider) ?? "CDN" : string.Empty;
+                    _MyObservStreamingEndpoints[index].CDN = ((bool)streamingEndpoint.Data.IsCdnEnabled.GetValueOrDefault()) ? StreamingEndpointInformation.ReturnDisplayedProvider(streamingEndpoint.Data.CdnProvider) ?? "CDN" : string.Empty;
```
same here

dima-ravnur commented on AMSExplorer/DataGrid/DataGridViewStreamingEndpoints.cs:
```diff
@@ -181,7 +181,7 @@ await foreach (var se in streamingendpointsPage)
                                 Id = c.Data.Id,
                                 Description = c.Data.Description,
-                                CDN = (bool)c.Data.IsCdnEnabled ? StreamingEndpointInformation.ReturnDisplayedProvider(c.Data.CdnProvider) ?? "CDN" : string.Empty,
+                                CDN = c.Data.IsCdnEnabled.GetValueOrDefault() ? StreamingEndpointInformation.ReturnDisplayedProvider(c.Data.CdnProvider) ?? "CDN" : string.Empty,
```
and here

dima-ravnur commented on AMSExplorer/Forms-Login/AMSLogin.cs:
```diff
@@ -234,11 +240,12 @@ private void ButtonExport_Click(object sender, EventArgs e)
 
                 PropertyRenameAndIgnoreSerializerContractResolver jsonResolver = new();
-                List<string> properties = new() { "EncryptedADSPClientSecret" };
+                List<string> properties = new() { "EncryptedADSPClientSecret", "MKIOEncryptedToken", "RavnurEncryptedApiKey" };
```
Why MKIOEncryptedToken was missing here? Doesn't it work without it?

vstratovych replied:
That seems to be a bug introduced when this property was added.

Later in the code there is a comment that describes intention not to export encrypted secrets.
Moreover, the property in CredentialsEntryV4 class contains comment that is very specific about ignoring that property during export.

dima-ravnur commented on AMSExplorer/AMSClient/AMSClientV3.cs:
```diff
@@ -194,20 +200,29 @@ public async Task<MediaServicesAccountResource> ConnectAndGetNewClientV3Async(Fo
+
+            if (IsRavnurClient)
+            {
+                armClient = RavnurClientFactory.GetRavnurArmClient(credentialsEntry);
```
Lets add some comment here that it is a key point where we make AMS SDK work with RMS API

dima-ravnur commented on AMSExplorer/Mainform.cs:
```diff
@@ -281,30 +287,72 @@ public Mainform(string[] args)
                 }
 
-                var leResults = _amsClient.AMSclient.GetMediaLiveEvents().GetAllAsync().ToListAsync().Result;
-                double nbLiveEvents = leResults.Count();
```
Don't we break existing functionality for AMS?

vstratovych replied:
No, I don't think so.
We just made sure that live events API isn't called for RMS.
For AMS everything should work just as it was before.

dima-ravnur commented on AMSExplorer/Mainform.cs:
```diff
@@ -7011,7 +7077,7 @@ private async Task ManageMenuOptionsAzureCDNAsync(ToolStripMenuItem disableAzure
                 var se = streamingendpoints.FirstOrDefault();
                 bool sestopped = (se.Data.ResourceState == StreamingEndpointResourceState.Stopped);
-                bool cdnenabled = (bool)se.Data.IsCdnEnabled;
+                bool cdnenabled = (bool)se.Data.IsCdnEnabled.GetValueOrDefault();
```
can we avoid it here as well?

vstratovych replied:
I'm going to revert this change and do fix in RMS API instead.

vstratovych committed:
Code clean up and minor improvements

Minor bug fixes and improvements
Use HLS as a default protocol for Ravnur links
Added check for *.json manifest in storage

vstratovych committed:
Version update and release notes

vstratovych committed:
Adjust UI for AMS only features

Disable controls related to features not available in RMS

vstratovych committed:
Open RMS links in Ravnur player by default

Replace links to Azure Media player with links to Ravnur player
when explorer connected to RMS instance

vstratovych committed:
Fix streaming link generation for RMS

Make sure proper streaming links are generated when
user want to play video in some player

vstratovych committed:
Fix streaming manifest generation for RMS
